sequence_processing_pipeline/ConvertJob.py

```diff
@@ -210,3 +218,158 @@ def parse_job_script(job_script_path):
+                      r"_S\d+_L\d+_[R,I]\d+_\d+.fastq.gz$"
+
+                for full_path in fastq_paths:
+                    file_path, file_name = split(full_path)
+                    if re.match(rgx, file_name):
+                        sample['matching_files'].append(full_path)
+
+        self.info = results
+
+    def copy_sequences(self, sample_name, source_project, dest_project,
+                       copy_all_replicates=False):
+        """
+        Copies all fastq files related to a sample into another project.
+        :param source_project: The source project w/qiita_id.
+        :param dest_project: The destination project w/qiita_id.
+        :param sample_name: A sample-name.
+        :param orig_name: A sample-name.
+        :param copy_all_replicates: If True, search for sample_name in the
+            orig_name column of the sample-sheet. Copy all replicates.
+        :return: None
+        """
+        if self.info is None:
+            raise ValueError("This method cannot be called until processing "
+                             "has completed.")
+
+        project_names = list(self.info.keys())
+
+        # confirm source project is a valid one.
+        if source_project not in project_names:
+            raise ValueError(f"'{source_project}' is not defined in the "
+                             "sample-sheet")
+
+        # confirm destination project is a valid one.
+        if dest_project not in project_names:
+            raise ValueError(f"'{dest_project}' is not defined in the "
+                             "sample-sheet")
+
+        if source_project == dest_project:
+            raise ValueError(f"source '{source_project}' and destination "
+                             f"'{dest_project}' projects are the same")
+
+        # note that the user can supply a sample-name that didn't make it
+        # through the conversion step and may have no files matched to it.
+        # this is considered okay and a possible outcome of conversion. In
+        # this case zero files are copied and no information is passed back
+        # to the user.
+
+        # projects that contain replicates must also be considered. if the
+        # value for sample_name contains a well-id, then only the files
+        # associated with that particular replicate should be copied. If the
+        # value instead references a sample_name in the 'orig_name' column,
+        # then all files associated with each replicate need to be moved.
+
+        # in this situation, the sample_name needs be compared against all
+        # orig_names in the project and the individual sample_names (w/well-
+        # ids) must be discovered. Then those individual sample_names can
+        # be processed.
+
+        if copy_all_replicates is True and self.contains_replicates is False:
+            raise ValueError("'treat_as_orig_name' is set to 'True' but this "
+                             "sample-sheet doesn't contain replicates")
```

Comment:
As a user of this method, I might well think, "Oh, sure, *if* this prep has any replicates, I want to go ahead and copy them, so I'll set `copy_all_replicates = True`!"  But if the prep in question didn't have replicates, this error condition would barf rather than just quietly saying, "well, there aren't any replicates so I don't have to worry about copying them!"  Can you clarify for me why it is undesirable to simply ignore the `copy_all_replicates` setting when a prep has no replicates?  I'm thinking something like 

```
name_col = "orig_name" if copy_all_replicates and self.contains_replicates else "Sample_Name"
for _, sample in samples:
   name_val = sample[name_col]
   if sample_name == name_val:
      sample_list.append(sample)
```

I acknowledge this would be theoretically slower, since it would loop over the samples in both cases rather than just the replicates case, but the logic would be simpler ...

Reply:
To be honest it's a question mark for me as well. :) Usually when I'm coding a new feature in SPP and I come up against a corner case that we didn't discuss, I'll do something explicit like this so that I can move forward and we can discuss it in the PR and change it if we decide to go another way.

I would be for simply ignoring it as well, but one question I had was - is there ever going to be a time when we wouldn't want to copy all replicates? if not then we can do away with the parameter entirely.

Reply:
Yes, your question is a really good one: is NOT copying all the replicates even something we *should* allow?  Based on my limited understanding of what replicates are in this context--the exact same extracted gDNA copied into multiple wells on a compression plate and library-prepped independently in each of those multiple wells--I cannot think of any legitimate reason for copying only SOME of them into a second project ... I would think it would be all or none.  In which case, maybe the user of this code should not have the option to set copy_all_replicates to False, and thus we can leave that parameter and its logic branch out.  I would be happy with that :)